gammapy/modeling/parameter.py

```diff
@@ -306,15 +306,19 @@ def scipy_mvn(self):
     def from_stack(cls, parameters_list):
         """Create `Parameters` by stacking smaller `Parameters`.
 
-        TODO: document
+        Parameters
+        ----------
+        parname : list
+            List of `Parameters`
         """
         pars = itertools.chain(*parameters_list)
-
-        # TODO: Fix covariance stacking!
-        # covariances = [Parameters(_)._any_covariance for _ in parameters_list]
-        # covariance = scipy.linalg.block_diag(*covariances)
-        covariance = None
-        return cls(pars, covariance)
+        parameters = cls(pars)
+        npars = len(parameters)
```

Comment:
I noticed that there is currently an inconsistency: here the covariance is set to `len(parameters)`, while in the function below, the indexing into the covariance matrix relies on `unique_parameters`. What is the convention currently used? Is the covariance adapted to the full or the unique parameter list. Can you please comment @cdeil?

Reply:
Always use the full parameters. Creating the subset of unique parameters should be an explicit operation, only used in the optimiser interface, never in internal methods.

Reply:
I'm just understanding the full consequences of the change to keep the full parameter list internally and I'm not very happy with it. Imaging a joint-fit of ~20 datasets and a model with ~20 parameter. So you get a 400x400 covariance matrix of which 95% of the information is duplicated? This does not makes sense to me. You also need to extend the API to things like `.unique_covariance` and `.set_covariance_factors()` to also extend the matrix for duplicated parameters. Currently the covariance is also not handled correctly in e.g. `_expand_factor_matrix()`, which still relies on the parameter list to be unique. It silently works now, because `Datasets.parameters` returns the unique parameter list. But the problem also appears here https://github.com/gammapy/gammapy/pull/2522/files#diff-c17632aab77905a57af61541e9dfb52bR118, where `.set_subcovariance()` relies on the unique parameter list to be passed in. My strong recommendation would be to undo the change to the full parameter list and filter out duplicated parameters in `.from_stack()` instead.